### PR TITLE
Support + tests for fit_generator + tensorboard

### LINKED ISSUE #1631: Tensorboard callback doesn't work on fit_generator
We have this error on a sequential model with multiple inputs:
keras sequential object as no attribute validation data

We have the same error for the graph:
keras graph object as no attribute validation data

It does work on fit !!

Is it normal?

`mix_model.fit_generator(trainBatchBuilder.seq_batch_generator(), 100, 5, validation_data=(X_test, Y_test),
                        callbacks=[TensorBoard("path/tensorboard_logs/", 1)])
`
Am I missing something?

### PR DESCRIPTION
**Pull request by tboquet**
This PR closes #1631.
The fit_generator method now adds the `validation_data` attribute to the model so that callbacks could access this data.

### CODE REVIEW

**Comment by tboquet** [comment]
I didn't check if my code was PEP8 compliant though. I forgot this feature was added in the build. I will take a look after the first build :).

**Comment by tboquet** [comment]
I'm adding more tests for the case when the user doesn't provide validation data and it's failing for now. I will work on this today.

**Comment by tboquet** [comment]
The `fit_generator` is now working with and without validation data with TensorBoard. The `validation_data` attribute is now added in both cases. PR #1647  is adding a fix but not addressing the missing attribute problem directly in the `fit_generator` method. Let me know if this work and if you think this PR solves the issue correctly @gw0, @fchollet, and I will squash all the commits.

File: keras/models.py [hunk 1]
```diff
@@ -972,6 +972,13 @@ def input_validation(generator_output):
                                 '2 or 3 elements.')
             return X, y, sample_weight
 
+        if do_validation:
+            X_val, y_val, sample_weight_val = input_validation(validation_data)
+            sample_weight_val = standardize_weights(y_val, sample_weight=sample_weight_val,
+                                                    sample_weight_mode=self.sample_weight_mode)
```
Comment (fchollet) [inline keras/models.py]: Sample weight standardization should be moved to `input_validation`. Same in `Graph`.

Comment (tboquet) [inline keras/models.py]: For `Graph`, I need the `sample_weight` to be a `dict` and I create a list to so it has the right format. I keep the dict so it could be passed to evaluate. Should I move the normalization of weights to `input_validation` and return a `list` and a `dict`?

File: keras/models.py [hunk 2]
```diff
@@ -1432,7 +1438,15 @@ def input_validation(generator_output):
                 raise Exception('All input arrays and target arrays must have '
                                 'the same number of samples.')
             return data, sample_weight
-
+        if do_validation:
+            data_val, sample_weight_val = input_validation(validation_data)
+            sample_weight_val_l = [standardize_weights(data_val[name],
```
Comment (tboquet) [inline keras/models.py]: For the `Graph`, I need the `sample_weight` to be a `dict` and I create a `list` so it has the right format for `self.validation_data`. I keep the `dict` so it could be passed to evaluate. Should I move the normalization of weights to `input_validation` and return a `list` and a `dict`?

Comment (fchollet) [inline keras/models.py]: You could standardize the weights in `input_validation`, return a dict, then turn the dict into a list afterwards. That way you still have the dict around for `evaluate`, but weight standardization is being done in just one place.

**Comment by tboquet** [comment]
The weight standardization has now been moved to the `input_validation` step.  Let me know if everything is alright for the `Graph.fit_generator` method.

**Comment by tboquet** [comment]
@fchollet I just revised the style, let me know if everything is ok and I'll squash the commits.

**Comment by fchollet** [comment]
The code looks reasonable to me, so we will merge. The tests look less reasonable: that's 9 different models to be compiled and trained, in order to test a really minor feature. That code would run every single time we run unit tests, which is on the order of 5-10 times a day. It's intense. It should be reduced to ~4 different models at most.

**Comment by tboquet** [comment]
Ok! tx very much for the review, I'm skimming the tests. The problem is that we were not able to refit the model because the `_set_model` method was merging summaries even if it was already done. I added a simple check to see if this step was completed. I'll commit the results soon, if everything goes as expected I should be able to propose a test with only 2 models compiled.

**Comment by tboquet** [comment]
Done :)! I was able to compile only 2 models because the accuracy is added as a value to the log. It now takes around 5s instead of 9s to run all the tests related to callbacks on my machine. I assume that no error means that TensorFlow can write to the log, not that the visualization will be correctly rendered by TensorBoard (because of the epochs alignment).

**Comment by fchollet** [comment]
This looks good to me. Please squash your commits and I will merge.

**Comment by tboquet** [comment]
@fchollet, done!
